### ISSUE TITLE
False positives for -Wunsequenced

### ISSUE THREAD

**Issue by Halalaluyafail3** [llvm/llvm-project#76546]
Clang interprets the evaluation of initializers as being unsequenced, which is incorrect.

For example the code here https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1729.htm in point B:

```c
#include <stdio.h>

#define ONE_INIT	'0' + i++ % 3
#define INITIALIZERS	[2] = ONE_INIT, [1] = ONE_INIT, [0] = ONE_INIT

int main()
{
	int i = 0;
	char x[4] = { INITIALIZERS }; // case 1
	puts(x);
	puts((char [4]){ INITIALIZERS }); // case 2
	puts((char [4]){ INITIALIZERS } + i % 2); // case 3
}
```

A warning is generated for case 3, which is correct; but no `-Wunsequenced` warnings should be generated for cases 1 or 2 which have defined behavior.

I found issue #32412 which references the same problem, and the conclusion reached there doesn't really make much sense.

> The evaluations of the initialization list expressions are indeterminately sequenced with respect to
> one another and thus the order in which any side effects occur is unspecified.

Section 6.7.9 "Initialization" Paragraph 23 ISO 9899:2018

> Evaluations A and B are indeterminately sequenced when A is
> sequenced either before or after B, but it is unspecified which.

Section 5.1.2.3 "Program execution" Paragraph 3 ISO 9899:2018

The evaluations of each initializer are sequenced and are in unspecified order, but this doesn't mean that they are unsequenced.

**Comment by dwblaikie** [comment]
> The evaluations of each initializer are sequenced and are in unspecified order, but this doesn't mean that they are unsequenced.

No, they aren't unsequenced, but they are indeterminately sequenced, and as @jyknight said ( https://github.com/llvm/llvm-project/issues/32412#issuecomment-980979266):

> So, these expressions should be considered "indeterminately-sequenced", not "un-sequenced". Note discussion in http://www.open-std.org/jtc1/sc22/wg14/www/docs/summary.htm#dr_439 which briefly touches upon this case as well, and also has that conclusion.
> 
> However, AFAICT, clang's -Wunsequenced warning is actually intended to warn about indeterminate-sequencing as well as strictly "unsequenced" expressions. Unless that's wrong, I think this is still working as intended.

**Comment by Halalaluyafail3** [comment]
> No, they aren't unsequenced, but they are indeterminately sequenced

I would expect the warning to indicate that they are indeterminately sequenced and not unsequenced, for example by creating a new warning -Windeterminately-sequenced.

**Comment by dwblaikie** [comment]
I wouldn't fundamentally object to some flag naming changes - though maybe a bit difficult to maintain backwards compatible with people expecting the existing warning under the existing flag. Maybe making `-Windeterminately-sequenced` a subgroup of `-Wunsequenced` (as much as that'd still be a bit confusing/bit of a misnomer)
